### PR TITLE
PIM-8709: pullup from 3.2 to master

### PR DESCRIPTION
**Pull request by olivierpontier**
**Description (for Contributor and Core Developer)**

This pull request aims to pull up all changes from 3.2 branch to the master

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added legacy Behats               | Todo
| Added acceptance tests            | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Bundle/Controller/InternalApi/ProductController.php [hunk 1]
```diff
@@ -112,6 +116,9 @@ class ProductController
      * @param NormalizerInterface           $constraintViolationNormalizer
      * @param ProductBuilderInterface       $variantProductBuilder
      * @param AttributeFilterInterface      $productAttributeFilter
+     * @param Client|null                   $productAndProductModelClient
+     *
+     * TODO: merge master remove null
```
Comment (olivierpontier) [inline src/Akeneo/Pim/Enrichment/Bundle/Controller/InternalApi/ProductController.php]: TODO

File: src/Akeneo/Pim/Enrichment/Bundle/Controller/InternalApi/ProductModelController.php [hunk 2]
```diff
@@ -105,6 +109,9 @@ class ProductModelController
      * @param NormalizerInterface               $violationNormalizer
      * @param FamilyVariantRepositoryInterface  $familyVariantRepository
      * @param AttributeFilterInterface          $productModelAttributeFilter
+     * @param Client|null                       $productAndProductModelClient
+     *
+     * TODO: merge master remove null
```
Comment (olivierpontier) [inline src/Akeneo/Pim/Enrichment/Bundle/Controller/InternalApi/ProductModelController.php]: TODO

File: src/Akeneo/Pim/Enrichment/Bundle/PdfGeneration/Renderer/ProductPdfRenderer.php [hunk 3]
```diff
@@ -55,6 +55,10 @@ class ProductPdfRenderer implements RendererInterface
     /** @var string */
     protected $customFont;
 
+    /** @var IdentifiableObjectRepositoryInterface|null */
+    private $attributeOptionRepository;
+
+    // TODO: on master we must remove null for the $attributeOptionRepository and change the order with $customFont
```
Comment (olivierpontier) [inline src/Akeneo/Pim/Enrichment/Bundle/PdfGeneration/Renderer/ProductPdfRenderer.php]: TODO

File: src/Akeneo/Pim/Enrichment/Component/Product/Normalizer/InternalApi/AxisValueLabelsNormalizer/MetricNormalizer.php [hunk 4]
```diff
@@ -4,14 +4,31 @@
...
 class MetricNormalizer implements AxisValueLabelsNormalizer
 {
+    /** @var StandardMetricNormalizer|null */
+    private $metricNormalizer;
+
+    /** @var MetricLocalizer|null */
+    private $metricLocalizer;
+
+    /**
+     * TODO: merge -> remove nullable and add BC break on UPGRADE
```
Comment (olivierpontier) [inline src/Akeneo/Pim/Enrichment/Component/Product/Normalizer/InternalApi/AxisValueLabelsNormalizer/MetricNormalizer.php]: TODO

File: src/Akeneo/Pim/Enrichment/Component/Product/Validator/Constraints/UniqueVariantAxisValidator.php [hunk 5]
```diff
@@ -33,19 +35,20 @@ class UniqueVariantAxisValidator extends ConstraintValidator
...
 
-    /**
-     * @param EntityWithFamilyVariantAttributesProvider  $axesProvider
-     * @param EntityWithFamilyVariantRepositoryInterface $repository
-     * @param UniqueAxesCombinationSet                   $uniqueAxesCombinationSet
-     */
+    /** @var GetValuesOfSiblings */
+    private $getValuesOfSiblings;
+
+    // TODO merge master/4.0: remove the second argument ($repository), and make the fourth one non nullable ($getValuesOfSiblings)
```
Comment (olivierpontier) [inline src/Akeneo/Pim/Enrichment/Component/Product/Validator/Constraints/UniqueVariantAxisValidator.php]: TODO

File: src/Akeneo/Pim/Enrichment/Component/Product/Validator/Constraints/UniqueVariantAxisValidator.php [hunk 6]
```diff
@@ -88,22 +91,26 @@ public function validate($entity, Constraint $constraint)
...
-        $siblings = $this->entityWithFamilyVariantRepository->findSiblings($entity);
+        $ownCombination = $this->getCombinationOfAxisValues($entity->getValuesForVariation(), $axes);
 
-        if (empty($siblings)) {
+        if ('' === str_replace(',', '', $ownCombination)) {
             return;
         }
 
-        $ownCombination = $this->getCombinationOfAxisValues($entity, $axes);
+        // TODO merge master/4.0: remove the test, and the whole 'else' statement
```
Comment (olivierpontier) [inline src/Akeneo/Pim/Enrichment/Component/Product/Validator/Constraints/UniqueVariantAxisValidator.php]: TODO

File: src/Akeneo/Pim/Enrichment/Component/Product/Validator/Constraints/UniqueVariantAxisValidator.php [hunk 7]
```diff
@@ -227,4 +234,15 @@ private function getEntityIdentifier(EntityWithFamilyVariantInterface $entity):
 
         return $entity->getCode();
     }
+
+    // TODO merge master/4.0 : remove this method
```
Comment (olivierpontier) [inline src/Akeneo/Pim/Enrichment/Component/Product/Validator/Constraints/UniqueVariantAxisValidator.php]: TODO

File: src/Akeneo/Platform/Bundle/UIBundle/Imagine/FlysystemLoader.php [hunk 8]
```diff
@@ -23,14 +24,15 @@ class FlysystemLoader implements LoaderInterface
...
 
-    /**
-     * @param FilesystemProvider $filesystemProvider
-     * @param array              $filesystemAliases
-     */
-    public function __construct(FilesystemProvider $filesystemProvider, array $filesystemAliases)
+    /** @var FileInfoRepositoryInterface | null */
+    protected $fileInfoRepository;
+
+    // TODO merge master/4.0: Make '$fileInfoRepository' mandatory.
```
Comment (olivierpontier) [inline src/Akeneo/Platform/Bundle/UIBundle/Imagine/FlysystemLoader.php]: TODO

File: src/Akeneo/Platform/Bundle/UIBundle/Imagine/FlysystemLoader.php [hunk 9]
```diff
@@ -85,6 +87,14 @@ protected function retrieveContentFileFromVfs($path)
             throw new NotLoadableException(sprintf('Unable to read the file "%s" from the filesystem.', $path));
         }
 
+        // TODO merge master/4.0: Remove 'null' check on '$this->fileInfoRepository'.
```
Comment (olivierpontier) [inline src/Akeneo/Platform/Bundle/UIBundle/Imagine/FlysystemLoader.php]: TODO

**Comment by olivierpontier** [comment]
Remove old changelogs?

File: src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductIdentifiers.php [hunk 10]
```diff
@@ -37,7 +39,11 @@ public function fetchByProductIdentifiers(array $productIdentifiers): array
     p.updated,
     f.code AS family_code,
     JSON_ARRAYAGG(g.code) AS group_codes,
-    JSON_MERGE(COALESCE(pm1.raw_values, '{}'), COALESCE(pm2.raw_values, '{}'), p.raw_values) as raw_values
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductIdentifiers.php]: Put back this line and drop the 4 nexts

File: src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductIdentifiers.php [hunk 11]
```diff
@@ -27,6 +27,8 @@ public function __construct(Connection $connection)
 
     public function fetchByProductIdentifiers(array $productIdentifiers): array
     {
+        // TODO TIP-1231: remove the 'IF' conditions in the JSON_MERGE_PRESERVE statement,
+        // and simply keep 'COALESCE(xxx.raw_values, '{}')
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductIdentifiers.php]: Remove comment

File: src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/ProductGrid/FetchProductModelRowsFromCodes.php [hunk 12]
```diff
@@ -248,7 +248,10 @@ private function getValueCollection(array $codes, array $attributeCodes, string
             SELECT 
                 pm.code,
                 a_label.code attribute_as_label_code,
-                JSON_MERGE(COALESCE(parent.raw_values, '{}'), pm.raw_values) as raw_values
+                JSON_MERGE(
+                    CASE parent.raw_values WHEN '[]' THEN '{}' ELSE COALESCE(parent.raw_values, '{}') END,
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/ProductGrid/FetchProductModelRowsFromCodes.php]: idem, put previous line as before

File: src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/ProductGrid/FetchProductRowsFromIdentifiers.php [hunk 13]
```diff
@@ -123,7 +123,11 @@ private function getValueCollection(array $identifiers, array $attributeCodes, s
                 p.identifier,
                 a_label.code attribute_as_label_code,
                 a_image.code attribute_as_image_code,
-                JSON_MERGE(COALESCE(pm1.raw_values, '{}'), COALESCE(pm2.raw_values, '{}'), p.raw_values) as raw_values
+                JSON_MERGE(
+                   CASE pm1.raw_values WHEN '[]' THEN '{}' ELSE COALESCE(pm1.raw_values, '{}') END,
+                   CASE pm2.raw_values WHEN '[]' THEN '{}' ELSE COALESCE(pm2.raw_values, '{}') END,
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/ProductGrid/FetchProductRowsFromIdentifiers.php]: idem

File: src/Akeneo/Pim/Structure/Bundle/Resources/config/datagrid/attribute.yml [hunk 14]
```diff
@@ -67,7 +67,7 @@ datagrid:
             columns:
                 code:
                     type:      string
-                    label:     pim_common.code
+                    label:     Code
```
Comment (pierallard) [inline src/Akeneo/Pim/Structure/Bundle/Resources/config/datagrid/attribute.yml]: put it back

File: src/Akeneo/Pim/Structure/Bundle/Resources/config/datagrid/family.yml [hunk 15]
```diff
@@ -51,7 +52,7 @@ datagrid:
             columns:
                 code:
                     type: string
-                    label: pim_common.code
+                    label: Code
```
Comment (pierallard) [inline src/Akeneo/Pim/Structure/Bundle/Resources/config/datagrid/family.yml]: put it back

File: src/Akeneo/Platform/Bundle/UIBundle/Resources/translations/jsmessages.ca_ES.yml [hunk 16]
```diff
@@ -121,7 +121,7 @@ pim_common:
   apply: Aplicar
   before: Abans
   by: per
-  cancel: Cancelar
+  cancel: Cancel·lar
```
Comment (nmarniesse) [inline src/Akeneo/Platform/Bundle/UIBundle/Resources/translations/jsmessages.ca_ES.yml]: Is it normal?

**Pull request merged**